### PR TITLE
[CALCITE-2446] Fix for lateral join in sql views

### PR DESCRIPTION
**Pull request by ptrbojko**

### CODE REVIEW

File: plus/src/main/java/org/apache/calcite/chinook/CodesFunction.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+      @Override public Queryable<String[]> asQueryable(QueryProvider queryProvider,
+                                                       SchemaPlus schema,
+                                                       String tableName) {
+        if (name == null) {
+          return Linq4j.asEnumerable(new String[][]{}).asQueryable();
+        }
+        return Linq4j.asEnumerable(new String[][]{
+            new String[]{"HASHCODE", "" + name.hashCode()},
+            new String[]{"BASE64",
+                Base64.encodeBase64String(name.getBytes(Charset.defaultCharset()))}
```
Comment (vlsi) [inline plus/src/main/java/org/apache/calcite/chinook/CodesFunction.java]: Please refrain from using `defaultCharset` in tests

File: plus/src/main/java/org/apache/calcite/chinook/CodesFunction.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+            .add("TYPE", SqlTypeName.VARCHAR)
+            .add("CODEVALUE", SqlTypeName.VARCHAR)
+            .build();
+      }
+
+      @Override public Queryable<String[]> asQueryable(QueryProvider queryProvider,
+                                                       SchemaPlus schema,
+                                                       String tableName) {
+        if (name == null) {
+          return Linq4j.asEnumerable(new String[][]{}).asQueryable();
```
Comment (vlsi) [inline plus/src/main/java/org/apache/calcite/chinook/CodesFunction.java]: Should this be `emptyEnumerable` ?

File: plus/pom.xml [hunk 3]
```diff
@@ -55,6 +55,11 @@ limitations under the License.
       <groupId>org.apache.calcite.avatica</groupId>
       <artifactId>avatica-core</artifactId>
     </dependency>
+    <dependency>
+      <groupId>commons-codec</groupId>
+      <artifactId>commons-codec</artifactId>
```
Comment (vlsi) [inline plus/pom.xml]: It does look more like a `<scope>test</scope>` dependency, doesn't it?
I would like to refrain from adding regular dependencies for test purposes.

File: core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java [hunk 4]
```diff
@@ -2049,7 +2049,7 @@ private SqlNode registerFrom(
               extendList,
               forceNullable,
               lateral);
-      if (newExpr != expr) {
+      if (newExpr != expr) { // && (isLateral(expr) && !isLateral(newExpr))
```
Comment (vlsi) [inline core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java]: I'm afraid this comment does not clarify much.

**Comment by ptrbojko** [comment]
@vlsi I've patched items from your review.

**Comment by zinking** [comment]
you may also want to squeeze the fix into 1 commit.

**Comment by vlsi** [comment]
@ptrbojko , I have no clue if `validator.setIdentifierExpansion(true);` is a proper approach or not.

**Comment by ptrbojko** [comment]
I've squashed the commits.
I've removed unneeded (as for the fix sake) - validator.setIdentifierExpansion(true); . There are view expansion methods in calcite and this change was a little step to make them look similar. But this wont fix the bug.

File: core/src/test/java/org/apache/calcite/util/Smalls.java [hunk 5]
```diff
@@ -126,6 +126,29 @@ public RelDataType getRowType(RelDataTypeFactory typeFactory) {
     };
   }
 
+  private static QueryableTable simpleTable(String s) {
```
Comment (vlsi) [inline core/src/test/java/org/apache/calcite/util/Smalls.java]: Is `simpleTable` vital for the test?
Can you reuse existing table function?

**Comment by ptrbojko** [comment]
@vlsi  I've reused existing tbl function instead of creating new one. As desired.

**Pull request closed without merging**